iproto: document `box.iproto` module (#3766)
tarantool/doc: doc/reference/reference_lua/box_iproto.rst

..  _box_iproto:

Submodule box.iproto
====================

Since :doc:`2.11.0 </release/2.11.0>`.

The ``box.iproto`` submodule provides the ability to work with the network subsystem of Tarantool.
It enables to extend the :ref:`binary protocol <box_protocol>` functionality from the Lua libraries.
With this submodule, you can:

*   parse the unknown IPROTO requests

> *   send arbitrary IPROTO packets

Reviewer: I would refer to the APIs this can be achieved with, i.e., `box.iproto.override` plus `IPROTO_UNKNOWN` for the former, and `box.iproto.send` for the latter.

> *   override the behavior of the existing request types in binary protocol

Reviewer: Ditto.

> To make this possible, the submodule exports all IPROTO :ref:`constants <box_iproto-constants>` and features to Lua.

Reviewer: I would refer to the IPROTO features [section](https://www.tarantool.io/ru/doc/latest/dev_guide/internals/iproto/keys/#iproto-features). 

> To make this possible

TBH, this is rather done for convenience. I would drop this line, as it doesn't provide useful information, IMHO.

..  _box_iproto-constants:

IPROTO constants
----------------

The IPROTO constants in the ``box.iproto`` namespace are written in the upper case without the ``IPROTO_`` prefix.

> The constants are divided into several types:

Reviewer: ```suggestion
The constants are divided into several groups:
```

*   :ref:`key <reference_lua-box_iproto_key>` (:ref:`IPROTO_SYNC <internals-iproto-keys-sync>`, :ref:`IPROTO_REQUEST_TYPE <internals-iproto-keys-request_type>`)
*   :ref:`request type <reference_lua-box_iproto_type>` (:ref:`IPROTO_OK <internals-iproto-ok>`)
*   :ref:`flag <reference_lua-box_iproto_flag>` (:ref:`IPROTO_COMMIT <box_protocol-commit>`)
*   :ref:`ballot key <reference_lua-box_iproto_ballot>` (:ref:`IPROTO_FLAG_COMMIT <box_protocol-flags>`)
*   :ref:`metadata key <reference_lua-box_iproto_metadata>` (:ref:`IPROTO_FIELD_IS_NULLABLE <internals-iproto-keys-sql-specific>`)
*   :ref:`RAFT key <reference_lua-box_iproto_raft>` (:ref:`IPROTO_TERM <internals-iproto-keys-term>`)

> Each type is located in the corresponding subnamespace without prefix.

Reviewer: ```suggestion
Each group is located in the corresponding subnamespace without prefix.
```

For example:

..  code-block:: lua

    box.iproto.key.SYNC = 0x01
    -- ...
    box.iproto.type.SELECT = 1
    -- ...
    box.iproto.flag.COMMIT = 0x01
    -- ...
    box.iproto.ballot_key.VCLOCK = 0x02
    -- ...
    box.iproto.metadata_key.IS_NULLABLE = 3
    -- ...
    box.iproto.raft_key.TERM = 0
    -- ...

..  _box_iproto-features:

IPROTO features
---------------

The submodule exports:

*   the current IPROTO protocol version (:ref:`box.iproto.protocol_version <reference_lua-box_iproto_version>`)
*   the set of IPROTO protocol features supported by the server (:ref:`box.iproto.protocol_features <reference_lua-box_iproto_protocol-features>`)

> *   IPROTO protocol features with the corresponding code (:ref:`box.iproto.feature <reference_lua-box_iproto_feature>`)

Reviewer: Maybe I would add a small example illustrating how this could be used (feel free to ignore this comment).

Author: Added an example

..  _box_iproto:

Submodule box.iproto
====================

Since :doc:`2.11.0 </release/2.11.0>`.

The ``box.iproto`` submodule provides the ability to work with the network subsystem of Tarantool.

> It enables to extend :ref:`IPROTO <box_protocol>` functionality from the Lua libraries.

Reviewer: ```suggestion
It allows to extend the :ref:`IPROTO <box_protocol>` functionality from Lua.
```

With this submodule, you can:

> *   :ref:`parse the unknown IPROTO requests types <reference_lua-box_iproto_override>`

Reviewer 2: ```suggestion
*   :ref:`parse unknown IPROTO request types <reference_lua-box_iproto_override>`
```

*   :ref:`send arbitrary IPROTO packets <reference_lua-box_iproto_send>`

> *   :ref:`override the behavior <reference_lua-box_iproto_override>` of the existing request types in the binary protocol

Reviewer: And unknown request types.

The submodule exports all IPROTO :ref:`constants <internals-box_protocol>` and :ref:`features <internals-iproto-keys-features>` to Lua.

..  _box_iproto-constants:

IPROTO constants
----------------

IPROTO constants in the ``box.iproto`` namespace are written in uppercase letters without the ``IPROTO_`` prefix.
The constants are divided into several groups:

*   :ref:`key <reference_lua-box_iproto_key>` (:ref:`IPROTO_SYNC <internals-iproto-keys-sync>`, :ref:`IPROTO_REQUEST_TYPE <internals-iproto-keys-request_type>`)

> *   :ref:`request type <reference_lua-box_iproto_type>` (:ref:`IPROTO_OK <internals-iproto-ok>`)

Reviewer 2: Am I correct that `IPROTO_OK` is one example of a request type? If yes, we can emphasise this somehow (`for example, IPROTO_OK`) and just give a link to the full list of keys.

*   :ref:`flag <reference_lua-box_iproto_flag>` (:ref:`IPROTO_COMMIT <box_protocol-commit>`)
*   :ref:`ballot key <reference_lua-box_iproto_ballot>` (:ref:`IPROTO_FLAG_COMMIT <box_protocol-flags>`)
*   :ref:`metadata key <reference_lua-box_iproto_metadata>` (:ref:`IPROTO_FIELD_IS_NULLABLE <internals-iproto-keys-sql-specific>`)

> *   :ref:`RAFT key <reference_lua-box_iproto_raft>` (:ref:`IPROTO_TERM <internals-iproto-keys-term>`)

Reviewer 2: As I see, this page already contains info, examples, and links to IPROTO internals for all namespaces listed in this list. I would think about dropping separate pages for API members (such as [box.iproto.key](https://docs.d.tarantool.io/en/doc/gh-3241-iproto/reference/reference_lua/box_iproto/key/)) and moving all these descriptions to the [API reference](https://docs.d.tarantool.io/en/doc/gh-3241-iproto/reference/reference_lua/box_iproto/#api-reference) section. But this up to you: I just like the [Every page is page one](https://idratherbewriting.com/trends/trends-to-follow-or-forget-every-page-is-page-one.html) approach  to give a user all the context at once :)

In a case you'll decide to merge all this, we can add separate subsections for namespaces, properties, and functions to `API reference` .

Each group is located in the corresponding subnamespace without the prefix.
For example:

..  code-block:: lua

    box.iproto.key.SYNC = 0x01
    -- ...
    box.iproto.type.SELECT = 1
    -- ...
    box.iproto.flag.COMMIT = 1
    -- ...
    box.iproto.ballot_key.VCLOCK = 2
    -- ...
    box.iproto.metadata_key.IS_NULLABLE = 3
    -- ...
    box.iproto.raft_key.TERM = 0
    -- ...

..  _box_iproto-features:

IPROTO features
---------------

The submodule exports:

*   the current IPROTO protocol version (:ref:`box.iproto.protocol_version <reference_lua-box_iproto_version>`)
*   the set of IPROTO protocol features supported by the server (:ref:`box.iproto.protocol_features <reference_lua-box_iproto_protocol-features>`)
*   IPROTO protocol features with the corresponding code (:ref:`box.iproto.feature <reference_lua-box_iproto_feature>`)

**Example**

The example converts the feature names from ``box.iproto.protocol_features`` set into codes:

..  code-block:: lua

    -- Features supported by the server
    box.iproto.protocol_features = {
    streams = true,
    transactions = true,
    error_extension = true,
    watchers = true,
    pagination = true,

>     }

Reviewer 2: Could you please reformat code to add indentation inside curly brackets?
<img width="550" alt="image" src="https://github.com/tarantool/doc/assets/38073144/b7e4ac92-baa9-44c1-9067-5a87c3d60c91">

:ref:`box.iproto.override() <reference_lua-box_iproto_override>` and :ref:`box_iproto_override <box_box_iproto_override>` functions.

..  _box_iproto-reference:

API reference
-------------

The table lists all available functions and data of the submodule:

..  container:: table

    ..  rst-class:: left-align-column-1
    ..  rst-class:: left-align-column-2

    ..  list-table::
        :widths: 30 70
        :header-rows: 1

        *   - Name
            - Use

        *   - :doc:`./box_iproto/key`
            - Request keys

        *   - :doc:`./box_iproto/request_type`
            - Request types

        *   - :doc:`./box_iproto/flag`
            - Flags from the :ref:`IPROTO_FLAGS <box_protocol-flags>` key

        *   - :doc:`./box_iproto/ballot`
            - Keys from the :ref:`IPROTO_BALLOT <box_protocol-ballots>` requests

        *   - :doc:`./box_iproto/metadata`
            - Keys nested in the :ref:`IPROTO_METADATA <internals-iproto-keys-metadata>` key

        *   - :doc:`./box_iproto/raft`
            - Keys from the ``IPROTO_RAFT_`` requests

        *   - :doc:`./box_iproto/protocol_version`

>             - Current IPROTO protocol version

Reviewer 2: ```suggestion
            - The current IPROTO protocol version
```
